chainer/functions/pooling_2d.py

```diff
@@ -352,3 +355,105 @@ def average_pooling_2d(x, ksize, stride=None, pad=0, use_cudnn=True):
 
     """
     return AveragePooling2D(ksize, stride, pad, False, use_cudnn)(x)
+
+
+class SpatialPyramidPooling2D(function.Function):
+
+    """Spatial pyramid pooling over a set of 2d planes."""
+
+    def __init__(self, x, pyramid_height, pooling_class, use_cudnn=True):
+        bottom_c, bottom_h, bottom_w = x.data.shape[1:]
+        assert bottom_h > 3 and bottom_w > 3, \
+            'input feature map size should be larger than 3'
+        self.pyramid_height = pyramid_height
+
+        # create pooling functions for different pyramid levels
+        split_inds = []
+        self.out_dim = 0
+        self.poolers = []
+        for pyramid_level in six.moves.range(pyramid_height):
+            num_bins = int(2 ** pyramid_level)
+
+            ksize_h = numpy.ceil(bottom_h / (float(num_bins)))
+            remainder_h = ksize_h * num_bins - bottom_h
+            pad_h = remainder_h / 2
+
+            ksize_w = numpy.ceil(bottom_w / (float(num_bins)))
+            remainder_w = ksize_w * num_bins - bottom_w
+            pad_w = remainder_w / 2
+
+            ksize = (int(ksize_h), int(ksize_w))
+            pad = (int(pad_h), int(pad_w))
+
+            if pooling_class == MaxPooling2D:
+                pooler = pooling_class(ksize=ksize, stride=None, pad=pad,
+                                       cover_all=True, use_cudnn=use_cudnn)
+                self.poolers.append(pooler)
+
+            if pooling_class == AveragePooling2D:
+                # cover_all mode is needed also for average pooling
+                raise NotImplementedError()
+
+            self.out_dim += bottom_c * (num_bins ** 2)
+            if pyramid_level < pyramid_height - 1:
+                split_inds.append(self.out_dim)
+
+        self.concat = concat.Concat(axis=1)
+        self.split = split_axis.SplitAxis(split_inds, axis=1)
+
+    def forward(self, x):
+        self.ys = []
+        for pooler in self.poolers:
+            y = pooler.forward(x)[0]
+            n, c, h, w = pooler.out_shape = y.shape
+            self.ys.append(y.reshape((n, c * h * w, 1, 1)))
+
+        return self.concat.forward(self.ys)
+
+    def backward_cpu(self, x, gy):
```

Comment:
`backward_cpu` and `backward_gpu` is almost the same. So it would be better to unify the two. There are (at least) two choice of implementation. One is to override `Function.backward` directly and the other is to make common routine a sub-function (like `_backward`). If we choose the first alternative, `backward_cpu` and `backward_gpu` will raise an `NotImplementedError`. But I think it is not problematic because some of existing functions do override `backward`.